[DOCS] Add notes on SSL version by JDK version (#73401)
elastic/elasticsearch: x-pack/docs/en/security/securing-communications/tls-versions-jdk.asciidoc

`SSLv3`::
  SSL v3 is supported on all JDKs, but is disabled by default.
  See <<jdk-enable-tls-protocol>>

`TLSv1`::
  TLS v1.0 is supported on all JDKs.
  Some newer JDKs, including the JDK bundled with {es} disable TLS v1.0 by default
  See <<jdk-enable-tls-protocol>>

`TLSv1.1`::
  TLS v1.1 is supported on all JDKs.
  Some newer JDKs, including the JDK bundled with {es} disable TLS v1.1 by default
  See <<jdk-enable-tls-protocol>>

`TLSv1.2`::
  TLS v1.2 is supported on all JDKs.
  It is enabled by default on all JDKs that are supported by {es}, including the bundled JDK.

`TLSv1.3`::
  TLS v1.3 is supported on all recent JDKs. This includes the latest release of Java 8
  (any release newer than 8u261) as well as all versions of Java 11 or later.
  It is supported and enabled by default on the {es} bundled JDK.

NOTE: The information above applies when running on the default JSSE provider in the JDK.
      JVMs that are configured to use a <<fips-140-compliance, FIPS 140-2>> security
      provider may have a custom TLS implementation and the supported TLS protocol versions
      may differ from the list above.
      +

>       Please check the release notes for your security provider.

Author: `for` sounds better than `of` to me. @lockewritesdocs ?

[[jdk-enable-tls-protocol]]
==== Enabling additional SSL/TLS versions on your JDK

The set of supported SSL/TLS versions for a JDK is controlled by a configuration file.
It is possible to override values that are set in this configuration file in order to

> enable legacy SSL or TLS versions.

Author: I'm worried that the tongue twister is hard to understand for people with limited english, (or for translation tools to handle).
I similarly tried to avoid using `re-enable`, but I'm happy to follow recommendations.

> enable legacy SSL or TLS versions.

Author: I've reworded this a bit. Is the new version clearer?

> enable legacy SSL or TLS versions.

Reviewer: Yep thanks

To do this, you should follow these steps:

1. Locate the configuration file for your JDK
2. Copy the `jdk.tls.disabledAlgorithms` setting from that file, and add it to a custom 
   configuration file within the {es} configuration directory
3. Modify the value of the `jdk.tls.disabledAlgorithms` setting so that the TLS version
   that you wish to use is no longer listed
4. Configure {es} to pass a custom system property to the JDK so that your custom
   configuration file is used.

===== Locate the configuration file for your JDK

For the {es} **bundled JDK**, the configuration file is in a sub directory of the {es} home directory (`$ES_HOME`).

* On Linux the configuration file is `$ES_HOME/jdk/conf/security/java.security` 
* On Windows the configuration file is `$ES_HOME/jdk/conf/security/java.security` 
* On macOS the configuration file is `$ES_HOME/jdk.app/Contents/Home/conf/security/java.security` 

For **JDK8**, the configuration file is within the `jre/lib/security` directory of the Java installation.
If `$JAVA_HOME` points to the home directory of the JDK that you use to run {es}, then the configuration file will be

* `$JAVA_HOME/jre/lib/security/java.security`

> For **JDK11 or later**, the configuration file is within the `conf/security` directory of the Java installation.

Reviewer: This does not include JDK9 and 10? They seem to match the pattern here (tested only with my local installations).

Author: Java 9 and 10 are EOL.
No supported version of ES is compatible with them.

    DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL
--------------------------------------------------

Copy this setting from the JDK's default configuration file, into a new file in your {es} configuration directory, named
`es.java.security`. You do not need to copy any other settings.

===== Enable required TLS versions

Edit the `es.java.security` file in your {es} configuration directory, and modify the `jdk.tls.disabledAlgorithms` setting
so that any SSL or TLS versions that you wish to use are no longer listed. 

For example, is you wished to enabled TLSv1.1 on OpenJDK (which uses the `jdk.tls.disabledAlgorithms` settings shown above),
then the contents of the `es.java.security` file would be:

[source,text]
--------------------------------------------------
jdk.tls.disabledAlgorithms=SSLv3, TLSv1, RC4, DES, MD5withRSA, \
    DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL
--------------------------------------------------

===== Enable your custom security configuration

Your custom security policy can be enabled by passing a system property to your JDK when you run {es}. This is achieved by

> adding a file into the <<set-jvm-options, `jvm.options.d`>> directory within your {es} configuration directory.

Author: We don't maintain docs for unsupported versions of Elasticsearch.
We may decide to write a version of these docs for 6.8 (which doesn't support jvm.options.d) but I'm not currently planning on backporting _this PR_ to `6.8`

[role="xpack"]
[[jdk-tls-versions]]
=== Supported SSL/TLS versions by JDK version

{es} relies on your JDK's implementation of SSL and TLS.

Different JDK versions support different versions of SSL, and this may affect how {es} operates.

`SSLv3`::
  SSL v3 is supported on all {es} <<jvm-version,compatible JDKs>> but is disabled by default.
  See <<jdk-enable-tls-protocol>>

`TLSv1`::
  TLS v1.0 is supported on all {es} <<jvm-version,compatible JDKs>>.
  Some newer JDKs, including the JDK bundled with {es} disable TLS v1.0 by default.
  See <<jdk-enable-tls-protocol>>

`TLSv1.1`::
  TLS v1.1 is supported on all {es} <<jvm-version,compatible JDKs>>.
  Some newer JDKs, including the JDK bundled with {es} disable TLS v1.1 by default.
  See <<jdk-enable-tls-protocol>>

`TLSv1.2`::
  TLS v1.2 is supported on all {es} <<jvm-version,compatible JDKs>>.
  It is enabled by default on all JDKs that are supported by {es}, including the bundled JDK.

`TLSv1.3`::

>   TLS v1.3 is supported on all recent {es} <<jvm-version,compatible JDKs>>.

Author: I was intentionally vague, because an in depth answer would take several paragraphs :)

What we mean is:
- TLS 1.3 is supported on JDK15 and JDK16.
- TLS 1.3 is supported on most releases of JDK11 (maybe all the GA ones, I'd need to check the history)
- TLS 1.3 is supported on some JDK8 builds, including the most recent release of each JDK8 distribution that ES supports
- However TLS 1.3 is not supported on _every_ JDK8 build, and ES supports running on older JDK8 builds that don't have TLS 1.3. We would encourage customers to upgrade to a JDK version that has TLS 1.3 (if only because newer JDKs are typically better), but it's not true to say "Every compatible JDK supports TLS 1.3"

If we can get that down to a couple of clear sentences that would be awesome.

   that you wish to use is no longer listed
4. Configure {es} to pass a custom system property to the JDK so that your custom
   configuration file is used.

===== Locate the configuration file for your JDK

For the {es} **bundled JDK**, the configuration file is in a sub directory of
the {es} home directory (`$ES_HOME`).

* On Linux the configuration file is `$ES_HOME/jdk/conf/security/java.security`
* On Windows the configuration file is `$ES_HOME/jdk/conf/security/java.security`
* On macOS the configuration file is `$ES_HOME/jdk.app/Contents/Home/conf/security/java.security`

For **JDK8**, the configuration file is within the `jre/lib/security` directory
of the Java installation.
If `$JAVA_HOME` points to the home directory of the JDK that you use to run {es},
then the configuration file will be

* `$JAVA_HOME/jre/lib/security/java.security`

For **JDK11 or later**, the configuration file is within the `conf/security`
directory of the Java installation.
If `$JAVA_HOME` points to the home directory of the JDK that you use to run
{es}, then the configuration file will be

* `$JAVA_HOME/conf/security/java.security`

===== Copy the disabledAlgorithms setting

> Within the configuration file (see above) there will be a line that starts with
> the text `jdk.tls.disabledAlgorithms=`

Reviewer 2: ```suggestion
Within the JDK configuration file is a line that starts with
`jdk.tls.disabledAlgorithms=`.
```

This setting controls which protocols and algorithms are _disabled_ in your JDK.
The value of that setting will typically span multiple lines.

For example, in OpenJDK 16 the setting is:
[source,text]
--------------------------------------------------
jdk.tls.disabledAlgorithms=SSLv3, TLSv1, TLSv1.1, RC4, DES, MD5withRSA, \
    DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL
--------------------------------------------------

> Copy this setting from the JDK's default configuration file, into a new file
> in your {es} configuration directory, named `es.java.security`.

Reviewer 2: ```suggestion
Create a new file in your in your {es} configuration directory named `es.java.security`.
Copy the `jdk.tls.disabledAlgorithms` setting from the JDK's default configuration file into `es.java.security`.
```

You do not need to copy any other settings.

===== Enable required TLS versions

Edit the `es.java.security` file in your {es} configuration directory, and
modify the `jdk.tls.disabledAlgorithms` setting so that any SSL or TLS versions
that you wish to use are no longer listed.

> For example, if you wished to enabled TLSv1.1 on OpenJDK (which uses the

Reviewer 2: ```suggestion
For example, if to enable TLSv1.1 on OpenJDK (which uses the
```

that you wish to use are no longer listed.

For example, if to enable TLSv1.1 on OpenJDK (which uses the
`jdk.tls.disabledAlgorithms` settings shown above), then the contents of the

> `es.java.security` file would include supported protocols _except_ `TLSv1.1`:

Reviewer: This sentence confuses me especially the "supported protocols" bit. I understand all the protocols are indeed supported. But it somehow jumps out for me and feels out of place. I think we can say something along the line of:
> then we would remove `TLSv1.1` from the file and the contents of the `es.java.secueity` file would become as follows:

Author: Thanks for the feedback - I agree.
I've tweaked the words slightly, can you have a look?

Reviewer: The new version is good. Thanks!

> `es.java.security` file would be:

Reviewer 2: ```suggestion
`es.java.security` file would include supported protocols _except_ `TLSv1.1`:
```

jdk.tls.disabledAlgorithms=SSLv3, TLSv1, RC4, DES, MD5withRSA, \
    DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL
--------------------------------------------------

===== Enable your custom security configuration

> Your custom security policy can be enabled by passing a system property to your
> JDK when you run {es}. This is achieved by adding a file into the
> <<set-jvm-options, `jvm.options.d`>> directory within your {es} configuration
> directory.

Reviewer 2: ```suggestion
To enable your custom security policy, add a file in the <<set-jvm-options, `jvm.options.d`>>
directory within your {es} configuration directory.
```

> 1. Locate the `jvm.options.d` directory for your {es} installation. See <<set-jvm-options>>

Reviewer 2: ```suggestion
1. Locate the `jvm.options.d` directory for your {es} installation.
```

Reviewer 2: Suggest removing that link since it's already in the intro.

> 2. Create a file named `java.security.options` within the `jvm.options.d` directory

Reviewer 2: ```suggestion
2. Create a file named `java.security.options` within the `jvm.options.d` directory.
```

> 3. The contents of that file should be set as per the example below.

Reviewer 2: ```suggestion
3. Set the path for the `es.java.security` file.
```

[source,text]

> --------------------------------------------------
> -Djava.security.properties=/path/to/your/es.java.security
> --------------------------------------------------

Reviewer 2: ```suggestion
----
-Djava.security.properties=/path/to/your/es.java.security
----
```

----
==== Enabling TLS versions in {es}

SSL/TLS versions can be enabled and disabled within {es} via the
<<ssl-tls-settings,`ssl.supported_protocols` settings>>.

{es} will only support the TLS versions that are enabled by the
<<jdk-tls-versions,underlying JDK>>. If you configure
`ssl.supported_procotols` to include a TLS version that is not enabled in your

> JDK, then it will be silently ignored.

Reviewer: Separately, I wonder whether it is possible to log some warnings for this situation.

{es}, then the configuration file will be in:

* `$JAVA_HOME/conf/security/java.security`

===== Copy the disabledAlgorithms setting

Within the JDK configuration file is a line that starts with
`jdk.tls.disabledAlgorithms=`.
This setting controls which protocols and algorithms are _disabled_ in your JDK.
The value of that setting will typically span multiple lines.

For example, in OpenJDK 16 the setting is:
[source,text]
--------------------------------------------------
jdk.tls.disabledAlgorithms=SSLv3, TLSv1, TLSv1.1, RC4, DES, MD5withRSA, \
    DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL
--------------------------------------------------

Create a new file in your in your {es} configuration directory named `es.java.security`.
Copy the `jdk.tls.disabledAlgorithms` setting from the JDK's default configuration file into `es.java.security`.
You do not need to copy any other settings.

===== Enable required TLS versions

Edit the `es.java.security` file in your {es} configuration directory, and
modify the `jdk.tls.disabledAlgorithms` setting so that any SSL or TLS versions
that you wish to use are no longer listed.

> For example, to enable TLSv1.1 on OpenJDK 16 (which uses the
> `jdk.tls.disabledAlgorithms` settings shown above), then the contents of the
> `es.java.security` file would contain the previously disabled TLS algorithms

Reviewer 2: ```suggestion
For example, to enable TLSv1.1 on OpenJDK 16 (which uses the
`jdk.tls.disabledAlgorithms` settings shown previously), the
`es.java.security` file would contain the previously disabled TLS algorithms
```

Reviewer 2: We try to avoid locational words like "above" and "below", which are not easily parsed by non-sighted users. Our docs use these words in many places, but it's best to avoid them if possible.

_except_ `TLSv1.1`:

[source,text]
--------------------------------------------------
jdk.tls.disabledAlgorithms=SSLv3, TLSv1, RC4, DES, MD5withRSA, \
    DH keySize < 1024, EC keySize < 224, 3DES_EDE_CBC, anon, NULL
--------------------------------------------------

===== Enable your custom security configuration

To enable your custom security policy, add a file in the <<set-jvm-options, `jvm.options.d`>>
directory within your {es} configuration directory.

1. Locate the `jvm.options.d` directory for your {es} installation.
2. Create a file named `java.security.options` within the `jvm.options.d` directory.

> 3. Set the path for the `es.java.security` file.

Reviewer 2: We can probably consolidate this information into two sentences:

> To enable your custom security policy, create a file named `java.security.options` within the <<set-jvm-options, `jvm.options.d`>> directory of your {es} configuration directory. Then, set the path for the `es.java.security` file from the command line:
